### PR TITLE
Two sided arnoldi

### PR DESCRIPTION
**Pull request by loiseaujc**
A relatively simple implementation of the two-sided Arnoldi factorization is implemented. It works out of the box although it could benefit eventually from some upgrades, most notably some restarting procedure.

### CODE REVIEW

**Comment by loiseaujc** [comment]
Note also that I moved all the `lapack` wrappers (i.e. `evd`, `hevd`, `svd`, `lstsq`) from `IterativeSolvers.f90` to `Utils.f90` for re-usability.

File: src/BaseKrylov.f90 [hunk 1]
```diff
@@ -461,4 +462,140 @@ subroutine nonsymmetric_lanczos_tridiagonalization(A, V, W, T, info, kstart, ken
...
+    k_start   = optval(kstart, 1)
+    k_end     = optval(kend, kdim)
+    verbose   = optval(verbosity, .false.)
+    tolerance = optval(tol, rtol)
+
+    !-----------------------------------------------
+    !-----     Two-sided Arnoldi iteration     -----
+    !-----------------------------------------------
+
+    arnoldi : do k = k_start, k_end
```
Comment (Simkern) [inline src/BaseKrylov.f90:529]: It seems you use the arnoldi routine that can do the full factorization inside a second loop to create the factorisations one step at a time. Is there a reason for this recursion? Why not do the following which seems clearer ?

```
call arnoldi_factorization(A, V, H, info, k_start, k_end, verbose, tolerance, .false.)
call arnoldi_factorization(A, W, G, info, k_start, k_end, verbose, tolerance, .false.)
```
I guess there is a good reason to do it this way. So maybe add a comment explaining it?

**Review by Simkern — COMMENTED** [review]
It looks good to me, there's only one step that seems convoluted to a new user like me and would benefit from a clarifying comment. Comment in code.

File: src/BaseKrylov.f90 (hunk 1 above)
Comment (loiseaujc) [reply src/BaseKrylov.f90:529]: Sure, `arnoldi_factorization` and all other algs in `BaseKrylov.f90` are implemented the same way.
It is essentially for convenience: the user can either perform a `k`-step Arnoldi factorization in one go, or a step-by-step one if they want compute some intermediate quantities, e.g. checking the number of converged eigenvalues and stop once a desired number has converged.

Basically:
- `call arnoldi_factorization(A, X, H, info)` will compute a full `k`-step factorization (where `k = size(X)-1` is the Krylov subspace dimension).
- Alternatively, you could have
  ```
  do k = 1, size(X)-1
      !> One step of Arnoldi fact.
      call arnoldi_factorization(A, X, H, info, k, k)
      !> Check eigenvalues for instance.
      if ( check(H) .is. .true.) then
          !> Export whatever you want.
          ...
          !> Exit loop.
          break
      endif
  enddo
  ```

Does that make sense? It basically is the same for `two_sided_arnoldi_factorization`. For eigenvalue computations, we want to stop the iteration once the desired number of eigentriplets have converged. For moment-matching reduced-order modeling, you want to do as many iterations as the rank of the low-order model you want to use. The same function can be used in both cases.

Comment (Simkern) [reply src/BaseKrylov.f90:529]: Yes, that makes sense! Thanks for the explanation!

**Review by Simkern — APPROVED** [review]

**Review by Simkern — COMMENTED** [review]
ok!

**Pull request merged**